zeshel/retrieval/data.py

```diff
@@ -0,0 +1,469 @@
+            random_cands_pool = list(range(len(self.all_entity_indices)))
+            random_cands_pool.remove(target_idx)
+            rand_cands = random.sample(random_cands_pool, self.num_rands)
+            negatives_token_ids = self.all_entity_token_ids[rand_cands].tolist()
+            candidate_token_ids += negatives_token_ids
+            negatives_masks = self.all_entity_masks[rand_cands].tolist()
+            candidate_masks += negatives_masks
+
+            # process hard negatives
+            if self.candidates is not None:
+                hard_negs = self.candidates[index]
+                hard_negative_token_ids = self.all_entity_token_ids[
+                    hard_negs].tolist()
+                hard_negative_masks = self.all_entity_masks[hard_negs].tolist()
+                candidate_token_ids += hard_negative_token_ids
+                candidate_masks += hard_negative_masks
+        elif self.type_cands == 'hard_adjusted_negative':
+            distributed_cands = self.candidates[0][index]
+            cands_probs = torch.tensor(self.candidates[1][index]).float()
+            negative_token_ids = self.all_entity_token_ids[
+                distributed_cands].tolist()
+            negative_masks = self.all_entity_masks[distributed_cands].tolist()
+            candidate_token_ids += negative_token_ids
+            candidate_masks += negative_masks
+        elif self.type_cands == 'hard_negative':
+            distributed_cands = self.candidates[index]
+            negative_token_ids = self.all_entity_token_ids[
+                distributed_cands].tolist()
+            negative_masks = self.all_entity_masks[distributed_cands].tolist()
+            candidate_token_ids += negative_token_ids
+            candidate_masks += negative_masks
+        else:
+            raise ValueError('wrong type candidates')
+
+        mention_token_ids = torch.tensor(mention_token_ids).long()
+        mention_masks = torch.tensor(mention_masks).long()
+        candidate_token_ids = torch.tensor(candidate_token_ids).long()
+        candidate_masks = torch.tensor(candidate_masks).long()
+        if self.type_cands == 'hard_adjusted_negative':
+            return mention_token_ids, mention_masks, candidate_token_ids, \
+                   candidate_masks, cands_probs
+        else:  # mixed/hard negative
+            return mention_token_ids, mention_masks, candidate_token_ids, \
+                   candidate_masks
+
+
+def transform_entities(doc, len_max, tokenizer):  # get all entities token
+    # ids and token masks
+    def get_entity_window(en_page_id):
+        page_id = en_page_id
+        en_title = doc[page_id]['title']
+        en_text = doc[page_id]['text']
+        max_len = len_max - 2  # cls , sep
+        ENT = '[unused2]'
+        en_title_tokens = tokenizer.tokenize(en_title)
+        en_text_tokens = tokenizer.tokenize(en_text)
+        window = (en_title_tokens + [ENT] + en_text_tokens)[:max_len]
```

Comment:
No need for a change, but just a comment that en_text_tokens already contains en_title_tokens (I think) so this formulation is redundant, and maybe skipping the title will improve performance because we see more text. Something to check in the future.